include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -95,100 +132,180 @@ namespace sparrow
-            throw std::runtime_error("Invalid ArrowArray format");
+            throw arrow_proxy_exception("Invalid ArrowArray format");
         }
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
+    template <typename AA, typename AS>
+    arrow_proxy::arrow_proxy(AA&& array, AS&& schema, bool)
         : m_array(std::move(array))
+        , m_array_ref(m_array.index() == 0 ? *std::get<0>(m_array) : std::get<1>(m_array))
         , m_schema(std::move(schema))
+        , m_schema_ref(m_schema.index() == 0 ? *std::get<0>(m_schema) : std::get<1>(m_schema))
     {
+        if constexpr (std::is_rvalue_reference_v<AA&&>)
+        {
+            array = {};
+        }
+        else if constexpr (std::is_pointer_v<std::remove_cvref_t<AA>>)
+        {
+            SPARROW_ASSERT_TRUE(array != nullptr);
+        }
+
+        if constexpr (std::is_rvalue_reference_v<AS&&>)
+        {
+            schema = {};
+        }
+        else if constexpr (std::is_pointer_v<std::remove_cvref_t<AS>>)
+        {
+            SPARROW_ASSERT_TRUE(schema != nullptr);
+        }
+
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
-        : m_array(std::move(array))
-        , m_schema(schema)
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
+        : arrow_proxy(std::move(array), std::move(schema), true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
+        : arrow_proxy(std::move(array), schema, true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(ArrowArray* array, ArrowSchema* schema)
+        : arrow_proxy(array, schema, true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(const arrow_proxy& other)
+        : m_array(deep_copy_array(other.m_array_ref, other.m_schema_ref))
+        , m_array_ref(std::get<1>(m_array))
+        , m_schema(deep_copy_schema(other.m_schema_ref))
+        , m_schema_ref(std::get<1>(m_schema))
     {
-        SPARROW_ASSERT_TRUE(schema != nullptr);
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray* array, ArrowSchema* schema)
-        : m_array(array)
-        , m_schema(schema)
+    inline arrow_proxy& arrow_proxy::operator=(const arrow_proxy& other)
     {
-        SPARROW_ASSERT_TRUE(array != nullptr);
-        SPARROW_ASSERT_TRUE(schema != nullptr);
+        if (this == &other)
+        {
+            return *this;
+        }
+
+        m_array = ArrowArray{};
+        m_array_ref = std::get<1>(m_array);
+        deep_copy_array(other.m_array_ref, other.m_schema_ref, m_array_ref);
+
+        m_schema = ArrowSchema{};
+        m_schema_ref = std::get<1>(m_schema);
+        deep_copy_schema(other.m_schema_ref, m_schema_ref);
```

Comment:
You have the same issue as in the move assign operator, the previous `m_array` and `m_schema` might never be released. A solution is to use the copy and swap idiom:

```cpp
        if (this == &other)
        {
            return *this;
        }

        arrow_proxy tmp(*other);
        swap(*this, tmp); // This assumes the existence of a swap function, you can also swap member by member
        return *this; 
```

Reply:
done